File: backend/app/engine/extreme_point.py
from typing import List, Tuple, Dict

ExtremePoint = Tuple[float, float, float]

SUPPORT_THRESHOLD = 0.1
VERTICAL_TOLERANCE = 1.0


def generate_new_eps(x: float, y: float, z: float,
                     item_l: float, item_h: float, item_w: float) -> List[ExtremePoint]:
    return [
        (x + item_l, y, z),
        (x, y + item_h, z),
        (x, y, z + item_w),
        (x + item_l, y, z + item_w),
        (x + item_l, y + item_h, z),
        (x, y + item_h, z + item_w),
    ]


def is_valid_ep(ep: ExtremePoint, item_size: Tuple[float, float, float],
                container: Tuple[float, float, float]) -> bool:
    ex, ey, ez = ep
    il, ih, iw = item_size
    cl, ch, cw = container
    return (ex + il <= cl + 0.001 and ey + ih <= ch + 0.001 and ez + iw <= cw + 0.001)


def check_overlap(x: float, y: float, z: float,
                  l: float, h: float, w: float,
                  placements: List[Dict]) -> bool:
    for p in placements:
        px, py, pz = p["x"], p["y"], p["z"]
        pl, ph, pw = p["l"], p["h"], p["w"]
        if (x < px + pl and x + l > px and
                y < py + ph and y + h > py and
                z < pz + pw and z + w > pz):
            return True
    return False


def check_support(x: float, y: float, z: float,
                  l: float, h: float, w: float,
                  placements: List[Dict]) -> Tuple[float, bool]:
    if y <= VERTICAL_TOLERANCE:
        return 1.0, True

    item_footprint = l * w
    if item_footprint <= 0:
        return 0.0, False

    supported_area = 0.0
    x2, z2 = x + l, z + w

    for p in placements:
        px, py, pz = p["x"], p["y"], p["z"]
        pl, ph, pw = p["l"], p["h"], p["w"]
        p_top = py + ph

        if abs(p_top - y) > VERTICAL_TOLERANCE:
            continue

        overlap_x = max(0, min(x2, px + pl) - max(x, px))
        overlap_z = max(0, min(z2, pz + pw) - max(z, pz))
        supported_area += overlap_x * overlap_z

    support_ratio = supported_area / item_footprint
    is_supported = support_ratio >= SUPPORT_THRESHOLD
    return round(support_ratio, 4), is_supported


def check_cg_stability(x: float, y: float, z: float,
                       l: float, h: float, w: float,
                       weight: float,
                       placements: List[Dict],
                       container: Tuple[float, float, float]) -> Tuple[float, bool]:
    total_weight = sum(p["weight"] for p in placements) + weight
    if total_weight <= 0:
        return 1.0, True

    new_cx = (sum(p["weight"] * (p["x"] + p["l"] / 2) for p in placements)
              + weight * (x + l / 2)) / total_weight
    new_cy = (sum(p["weight"] * (p["y"] + p["h"] / 2) for p in placements)
              + weight * (y + h / 2)) / total_weight
    new_cz = (sum(p["weight"] * (p["z"] + p["w"] / 2) for p in placements)
              + weight * (z + w / 2)) / total_weight

    cl, ch, cw = container

    cx_center = cl / 2
    cz_center = cw / 2

    allowed_offset_x = cl * 0.35
    allowed_offset_z = cw * 0.35

    offset_x = abs(new_cx - cx_center)
    offset_z = abs(new_cz - cz_center)

    cg_score = max(0.0, 1.0 - (offset_x / allowed_offset_x + offset_z / allowed_offset_z) / 2)
    height_penalty = 0.0
    if ch > 0:
        height_penalty = min(1.0, new_cy / (ch * 0.5))
    cg_score = cg_score * (1.0 - height_penalty * 0.3)

    is_stable = (offset_x <= allowed_offset_x and offset_z <= allowed_offset_z)

    return round(cg_score, 4), is_stable


def check_fragile_safety(x: float, y: float, z: float,
                         l: float, h: float, w: float,
                         weight: float,
                         placements: List[Dict]) -> Tuple[float, bool]:
    x2, y2, z2 = x + l, y + h, z + w
    penalty = 0.0
    has_issue = False

    for p in placements:
        if not p.get("is_fragile"):
            continue
        px, py, pz = p["x"], p["y"], p["z"]
        pl, ph, pw = p["l"], p["h"], p["w"]
        p_top = py + ph

        if abs(y - p_top) > VERTICAL_TOLERANCE:
            continue

        overlap_x = max(0, min(x2, px + pl) - max(x, px))
        overlap_z = max(0, min(z2, pz + pw) - max(z, pz))
        overlap_area = overlap_x * overlap_z
        fragile_area = pl * pw

        if overlap_area > 0 and fragile_area > 0:
            coverage = overlap_area / fragile_area
            if coverage > 0.1:
                penalty += coverage * weight
                has_issue = True

    safety_score = max(0.0, 1.0 - penalty)
    return round(safety_score, 4), not has_issue


def evaluate_placement(ep: ExtremePoint, item_size: Tuple[float, float, float],
                       container: Tuple[float, float, float],
                       existing_placements: List[Dict],
                       weight: float = 1.0,
                       is_fragile: bool = False) -> Tuple[float, Dict]:
    """评估放置位置得分（空间利用率最优版本）。

    优化点：
    - 移除 O(N) 的 cg_score / gravity_score / fragile_safe / adj_score 计算
    - 仅保留 O(1) 的 BLF（Bottom-Left-Fill）评分
    - 不考虑重量平衡和稳定性约束
    """
    ex, ey, ez = ep
    il, ih, iw = item_size
    cl, ch, cw = container

    # BLF 评分：优先填底部、左部、后部，O(1) 复杂度
    if cl > 0 and ch > 0 and cw > 0:
        blf_score = 1.0 - (ey / ch * 0.5 + ex / cl * 0.3 + ez / cw * 0.2)
        blf_score = max(0.0, min(1.0, blf_score))
    else:
        blf_score = 1.0

    # 紧贴容器壁奖励：物品贴边时给予小幅加分，鼓励规则堆叠
    wall_bonus = 0.0
    if ex <= 0.01 or ex + il >= cl - 0.01:
        wall_bonus += 0.05
    if ez <= 0.01 or ez + iw >= cw - 0.01:
        wall_bonus += 0.05
    if ey <= 0.01:
        wall_bonus += 0.05

    total_score = blf_score + wall_bonus

    metrics = {
        "blf_score": round(blf_score, 4),
        "wall_bonus": round(wall_bonus, 4),
    }

    return round(total_score, 4), metrics


def evaluate_placement_fast(ep: ExtremePoint, item_size: Tuple[float, float, float],
                             container: Tuple[float, float, float],
                             existing_placements: List[Dict],
                             weight: float = 1.0,
                             is_fragile: bool = False) -> float:
    """快速评估放置位置得分（与 evaluate_placement 一致，O(1) 复杂度）。"""
    ex, ey, ez = ep
    il, ih, iw = item_size
    cl, ch, cw = container

    if cl > 0 and ch > 0 and cw > 0:
        blf_score = 1.0 - (ey / ch * 0.5 + ex / cl * 0.3 + ez / cw * 0.2)
        blf_score = max(0.0, min(1.0, blf_score))
    else:
        blf_score = 1.0

    wall_bonus = 0.0
    if ex <= 0.01 or ex + il >= cl - 0.01:
        wall_bonus += 0.05
    if ez <= 0.01 or ez + iw >= cw - 0.01:
        wall_bonus += 0.05
    if ey <= 0.01:
        wall_bonus += 0.05

    return round(blf_score + wall_bonus, 4)

File: src-tauri/src/extreme_point.rs
/// Extreme Point 工具函数模块

use crate::models::PlacementInfo;

pub const SUPPORT_THRESHOLD_EP: f64 = 0.1;
pub const VERTICAL_TOLERANCE: f64 = 1.0;

pub fn generate_new_eps(
    x: f64, y: f64, z: f64,
    item_l: f64, item_h: f64, item_w: f64,
) -> Vec<(f64, f64, f64)> {
    vec![
        (x + item_l, y, z),
        (x, y + item_h, z),
        (x, y, z + item_w),
        (x + item_l, y, z + item_w),
        (x + item_l, y + item_h, z),
        (x, y + item_h, z + item_w),
    ]
}

pub fn is_valid_ep(
    ep: (f64, f64, f64),
    item_size: (f64, f64, f64),
    container: (f64, f64, f64),
) -> bool {
    let (ex, ey, ez) = ep;
    let (il, ih, iw) = item_size;
    let (cl, ch, cw) = container;
    ex + il <= cl + 0.001 && ey + ih <= ch + 0.001 && ez + iw <= cw + 0.001
}

pub fn check_overlap(
    x: f64, y: f64, z: f64,
    l: f64, h: f64, w: f64,
    placements: &[PlacementInfo],
) -> bool {
    for p in placements {
        if x < p.x + p.l
            && x + l > p.x
            && y < p.y + p.h
            && y + h > p.y
            && z < p.z + p.w
            && z + w > p.z
        {
            return true;
        }
    }
    false
}

pub fn check_support(
    x: f64, y: f64, z: f64,
    l: f64, h: f64, w: f64,
    placements: &[PlacementInfo],
) -> (f64, bool) {
    if y <= VERTICAL_TOLERANCE {
        return (1.0, true);
    }
    let item_footprint = l * w;
    if item_footprint <= 0.0 {
        return (0.0, false);
    }
    let mut supported_area = 0.0;
    let x2 = x + l;
    let z2 = z + w;

    for p in placements {
        let p_top = p.y + p.h;
        if (p_top - y).abs() > VERTICAL_TOLERANCE {
            continue;
        }
        let overlap_x = (x2.min(p.x + p.l) - x.max(p.x)).max(0.0);
        let overlap_z = (z2.min(p.z + p.w) - z.max(p.z)).max(0.0);
        supported_area += overlap_x * overlap_z;
    }

    let support_ratio = supported_area / item_footprint;
    let is_supported = support_ratio >= SUPPORT_THRESHOLD_EP;
    ((support_ratio * 10000.0).round() / 10000.0, is_supported)
}

pub fn check_cg_stability(
    x: f64, y: f64, z: f64,
    l: f64, h: f64, w: f64,
    weight: f64,
    placements: &[PlacementInfo],
    container: (f64, f64, f64),
) -> (f64, bool) {
    let total_weight: f64 = placements.iter().map(|p| p.weight).sum::<f64>() + weight;
    if total_weight <= 0.0 {
        return (1.0, true);
    }
    let new_cx = (placements.iter().map(|p| p.weight * (p.x + p.l / 2.0)).sum::<f64>()
        + weight * (x + l / 2.0))
        / total_weight;
    let new_cy = (placements.iter().map(|p| p.weight * (p.y + p.h / 2.0)).sum::<f64>()
        + weight * (y + h / 2.0))
        / total_weight;
    let new_cz = (placements.iter().map(|p| p.weight * (p.z + p.w / 2.0)).sum::<f64>()
        + weight * (z + w / 2.0))
        / total_weight;

    let (cl, ch, cw) = container;
    let cx_center = cl / 2.0;
    let cz_center = cw / 2.0;
    let allowed_offset_x = cl * 0.35;
    let allowed_offset_z = cw * 0.35;
    let offset_x = (new_cx - cx_center).abs();
    let offset_z = (new_cz - cz_center).abs();

    let mut cg_score = (1.0 - (offset_x / allowed_offset_x + offset_z / allowed_offset_z) / 2.0).max(0.0);
    let height_penalty = if ch > 0.0 {
        (new_cy / (ch * 0.5)).min(1.0)
    } else {
        0.0
    };
    cg_score *= 1.0 - height_penalty * 0.3;

    let is_stable = offset_x <= allowed_offset_x && offset_z <= allowed_offset_z;
    ((cg_score * 10000.0).round() / 10000.0, is_stable)
}

pub fn check_fragile_safety(
    x: f64, y: f64, z: f64,
    l: f64, h: f64, w: f64,
    weight: f64,
    placements: &[PlacementInfo],
) -> (f64, bool) {
    let x2 = x + l;
    let z2 = z + w;
    let mut penalty = 0.0;
    let mut has_issue = false;

    for p in placements {
        if !p.is_fragile {
            continue;
        }
        let p_top = p.y + p.h;
        if (y - p_top).abs() > VERTICAL_TOLERANCE {
            continue;
        }
        let overlap_x = (x2.min(p.x + p.l) - x.max(p.x)).max(0.0);
        let overlap_z = (z2.min(p.z + p.w) - z.max(p.z)).max(0.0);
        let overlap_area = overlap_x * overlap_z;
        let fragile_area = p.l * p.w;
        if overlap_area > 0.0 && fragile_area > 0.0 {
            let coverage = overlap_area / fragile_area;
            if coverage > 0.1 {
                penalty += coverage * weight;
                has_issue = true;
            }
        }
    }

    let safety_score = (1.0 - penalty).max(0.0);
    ((safety_score * 10000.0).round() / 10000.0, !has_issue)
}

pub fn evaluate_placement(
    ep: (f64, f64, f64),
    item_size: (f64, f64, f64),
    container: (f64, f64, f64),
    _existing_placements: &[PlacementInfo],
    _weight: f64,
    _is_fragile: bool,
) -> (f64, std::collections::HashMap<String, f64>) {
    let (ex, ey, ez) = ep;
    let (il, ih, iw) = item_size;
    let (cl, ch, cw) = container;

    let blf_score = if cl > 0.0 && ch > 0.0 && cw > 0.0 {
        let s = 1.0 - (ey / ch * 0.5 + ex / cl * 0.3 + ez / cw * 0.2);
        s.max(0.0).min(1.0)
    } else {
        1.0
    };

    let mut wall_bonus = 0.0;
    if ex <= 0.01 || ex + il >= cl - 0.01 { wall_bonus += 0.05; }
    if ez <= 0.01 || ez + iw >= cw - 0.01 { wall_bonus += 0.05; }
    if ey <= 0.01 { wall_bonus += 0.05; }

    let total_score = blf_score + wall_bonus;
    let mut metrics = std::collections::HashMap::new();
    metrics.insert("blf_score".to_string(), (blf_score * 10000.0).round() / 10000.0);
    metrics.insert("wall_bonus".to_string(), (wall_bonus * 10000.0).round() / 10000.0);
    ((total_score * 10000.0).round() / 10000.0, metrics)
}

pub fn evaluate_placement_fast(
    ep: (f64, f64, f64),
    item_size: (f64, f64, f64),
    container: (f64, f64, f64),
    _existing_placements: &[PlacementInfo],
    _weight: f64,
    _is_fragile: bool,
) -> f64 {
    let (ex, ey, ez) = ep;
    let (il, ih, iw) = item_size;
    let (cl, ch, cw) = container;

    let blf_score = if cl > 0.0 && ch > 0.0 && cw > 0.0 {
        let s = 1.0 - (ey / ch * 0.5 + ex / cl * 0.3 + ez / cw * 0.2);
        s.max(0.0).min(1.0)
    } else {
        1.0
    };

    let mut wall_bonus = 0.0;
    if ex <= 0.01 || ex + il >= cl - 0.01 { wall_bonus += 0.05; }
    if ez <= 0.01 || ez + iw >= cw - 0.01 { wall_bonus += 0.05; }
    if ey <= 0.01 { wall_bonus += 0.05; }

    ((blf_score + wall_bonus) * 10000.0).round() / 10000.0
}
